File: tests/fixtures/rule_violations/java/BadPatterns.java
// Java file with intentional rule violations for testing.
//
// Each violation is annotated with // expect: <rule-id> on the same line.
package com.example;

import java.util.List;

public class BadPatterns {

    public String buildMessage(List<String> items) {
        // Builtin: string concat in loop
        String result = "";
        for (String item : items) {
            result += "item: " + item; // expect: java-string-concat-loop
        }
        System.out.println("Done: " + result); // expect: java-no-system-out
        return result;
    }

    public void riskyMethod() {
        try {
            loadData();
        } catch (Exception e) { // expect: java-catch-exception
            System.out.println(e.getMessage()); // expect: java-no-system-out
        }
    }

    private void cleanLoadData() throws Exception {
        // clean — no findings expected
        throw new Exception("not implemented");
    }
}

File: tests/fixtures/rule_violations/python/bad_patterns.py
"""Python file with intentional rule violations for testing.

Each violation is annotated with # expect: <rule-id> on the same line.
The test runner verifies every expect annotation produces a finding,
and no unexpected findings appear.
"""
import logging
from utils import *  # expect: no-star-import

logger = logging.getLogger(__name__)


def process_items(items: list, config: dict) -> None:
    """Process items with several bad patterns."""
    # Team preference: no print debugging
    print(f"Starting with {len(items)} items")  # expect: no-print-debugging

    # Team preference: no bare dict access
    api_key = config['api_key']  # expect: no-bare-dict-access

    # Team preference: no f-string in logger
    logger.info(f"Using key {api_key[:4]}...")  # expect: no-string-format-logging

    for item in items:
        print(item.name)  # expect: no-print-debugging
        status = config['default_status']  # expect: no-bare-dict-access


def clean_calculate_total(prices: list[float]) -> float:
    """Clean function — should produce NO findings."""
    total = sum(prices)
    logger.debug("Calculated total: %s", total)
    return total


def clean_load_config(path: str) -> dict:
    """Another clean function."""
    with open(path) as f:
        return {}
